cli/upload/upload.go

```diff
@@ -53,7 +54,8 @@ func NewCommand() *cobra.Command {
 
 	uploadCommand.Flags().StringVarP(&fqbn, "fqbn", "b", "", "Fully Qualified Board Name, e.g.: arduino:avr:uno")
 	uploadCommand.Flags().StringVarP(&port, "port", "p", "", "Upload port, e.g.: COM10 or /dev/ttyACM0")
-	uploadCommand.Flags().StringVarP(&importDir, "input-dir", "", "", "Direcory containing binaries to upload.")
+	uploadCommand.Flags().StringVarP(&importDir, "input-dir", "", "", "Directory containing binaries to upload.")
+	uploadCommand.Flags().StringVarP(&importFile, "input-file", "", "", "Binary file to upload.")
```

Comment:
What happens if `--input-file` and `--input-dir` are both set?

I think we must add an error for this particular case and we must explicitly say that `--input-file` and `sketch --input-dir buildpath/...` are mutually exclusive.

Reply:
I mean we have 8 cases:

`arduino-cli upload` -> OK (sketch upload on working directory)

`arduino-cli upload Blink` -> OK (sketch upload)

`arduino-cli upload --input-file Blink.ino` -> OK (single file upload NO sketch)

`arduino-cli upload Blink --input-file Blink.ino`
error "--input-file cannot be used with a sketch, use --input-dir instead, or remove the sketch path"

`arduino-cli upload --input-path /path/to/Blink/build` -> OK (sketch upload on working directory using external build path)

`arduino-cli upload Blink --input-path /path/to/Blink/build` -> OK (sketch upload using external build path)

`arduino-cli upload --input-path /path/to/Blink/build --input-file some/other/Blink.ino`
error "--input-file and --input-dir cannot be used together"

`arduino-cli upload Blink --input-path /path/to/Blink/build --input-file some/other/Blink.ino`
error "--input-file and --input-dir cannot be used together"

Reply:
This whole process and flow needs some revisiting in order to make it solid yet flexible and confusion free.
I can see long days ahead :D

Reply:
See e10d0ff,
I used the [PreRun Cobra facility](https://github.com/spf13/cobra#prerun-and-postrun-hooks) to check at CLI level the presence of both flags, avoiding to add messy `if`s to the the `commands/upload.go` internals.